File: src/terminal/runLifecycle.ts
import type { CliKind, U64String } from '@/types/cli'
import type { OutputState, ProcessState, RunState } from '@/types/terminal'

const MAX_U64 = (1n << 64n) - 1n

export type RunLifecycleEvent =
  | { type: 'process-exited'; runId: string; generation: number }
  | { type: 'output-end'; runId: string; generation: number; streamEpoch: U64String; finalOffset: U64String }
  | { type: 'degraded'; runId: string; generation: number; streamEpoch: U64String }
  | { type: 'incomplete'; runId: string; generation: number; streamEpoch: U64String }

export interface RunLifecycleSnapshot extends RunState {
  streamEpoch?: U64String
  finalOffset?: U64String
  parsedThrough: U64String
  sentThrough: U64String
}

export interface RunLifecycle {
  snapshot(): RunLifecycleSnapshot
  accept(event: RunLifecycleEvent): boolean
  processRunning(): void
  processExited(): void
  outputStarted(streamEpoch: U64String): void
  sentThrough(offset: U64String): void
  outputEnd(finalOffset: U64String): void
  parsedThrough(offset: U64String): void
  degraded(): void
  incomplete(): void
}

function parseU64(value: string): bigint {
  if (!/^(0|[1-9][0-9]*)$/.test(value)) throw new Error('INVALID_U64')
  const parsed = BigInt(value)
  if (parsed > MAX_U64) throw new Error('INVALID_U64')
  return parsed
}

export function createRunLifecycle(input: {
  runId: string
  generation: number
  cli: CliKind
}): RunLifecycle {
  let process: ProcessState = 'starting'
  let output: OutputState = 'open'
  let streamEpoch: U64String | undefined
  let sent = 0n
  let parsed = 0n
  let finalOffset: bigint | undefined

  const refresh = () => {
    if (output === 'degraded' || output === 'incomplete') return
    if (finalOffset !== undefined && parsed === finalOffset) output = 'drained'
    else if (finalOffset !== undefined || process === 'exited') output = 'draining'
  }

  const requireEpoch = (): U64String => {
    if (!streamEpoch) throw new Error('OUTPUT_STREAM_NOT_READY')
    return streamEpoch
  }

  const lifecycle: RunLifecycle = {
    snapshot() {
      return {
        runId: input.runId,
        generation: input.generation,
        cli: input.cli,
        process,
        output,
        activity: 'unknown',
        observation: 'off',
        ...(streamEpoch ? { streamEpoch } : {}),
        ...(finalOffset !== undefined ? { finalOffset: finalOffset.toString() as U64String } : {}),
        parsedThrough: parsed.toString() as U64String,
        sentThrough: sent.toString() as U64String,
      }
    },

    accept(event) {
      if (event.runId !== input.runId || event.generation !== input.generation) return false
      if ('streamEpoch' in event && streamEpoch !== event.streamEpoch) return false
      switch (event.type) {
        case 'process-exited':
          lifecycle.processExited()
          return true
        case 'output-end':
          lifecycle.outputEnd(event.finalOffset)
          return true
        case 'degraded':
          lifecycle.degraded()
          return true
        case 'incomplete':
          lifecycle.incomplete()
          return true
      }
    },

    processRunning() {
      if (process !== 'starting' && process !== 'running') throw new Error('RUN_STATE_CONFLICT')
      process = 'running'
    },

    processExited() {
      if (process === 'failed') throw new Error('RUN_STATE_CONFLICT')
      process = 'exited'
      refresh()
    },

    outputStarted(epoch) {
      if (parseU64(epoch) === 0n) throw new Error('INVALID_STREAM_EPOCH')
      if (streamEpoch && streamEpoch !== epoch) throw new Error('STALE_OUTPUT_STREAM')
      streamEpoch = epoch
    },

    sentThrough(value) {
      requireEpoch()
      const next = parseU64(value)
      if (next < sent) throw new Error('OUTPUT_OFFSET_BACKWARD')
      if (finalOffset !== undefined && next !== sent) throw new Error('OUTPUT_ALREADY_ENDED')
      sent = next
    },

    outputEnd(value) {
      requireEpoch()
      const next = parseU64(value)
      if (next > sent) throw new Error('OUTPUT_END_BEYOND_SENT')
      if (finalOffset !== undefined && finalOffset !== next) throw new Error('OUTPUT_END_CONFLICT')
      finalOffset = next
      refresh()
    },

    parsedThrough(value) {
      requireEpoch()
      const next = parseU64(value)
      if (next < parsed) throw new Error('OUTPUT_ACK_BACKWARD')
      if (next > sent) throw new Error('OUTPUT_ACK_BEYOND_SENT')
      parsed = next
      refresh()
    },

    degraded() {
      requireEpoch()
      output = 'degraded'
    },

    incomplete() {
      requireEpoch()
      output = 'incomplete'
    },
  }

  return lifecycle
}

File: src-tauri/src/run_lifecycle.rs
//! Process/output lifecycle state independent from PTY transport ownership.
#![allow(dead_code)] // D15 supervisor wiring follows this state-machine checkpoint.

use crate::cli::profiles::error;
use crate::cli::run_registry::RunKey;
use crate::cli::types::{SafeError, WireU64};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum ProcessLifecycle {
    Starting,
    Running,
    Exited,
    Failed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum OutputLifecycle {
    Open,
    Draining,
    Drained,
    Incomplete,
    Degraded,
}

#[derive(Debug, Clone)]
pub(crate) struct LifecycleRecord {
    run: RunKey,
    process: ProcessLifecycle,
    output: OutputLifecycle,
    stream_epoch: Option<WireU64>,
    sent: u64,
    parsed: u64,
    final_offset: Option<u64>,
}

impl LifecycleRecord {
    pub(crate) fn new(run: RunKey) -> Self {
        Self {
            run,
            process: ProcessLifecycle::Starting,
            output: OutputLifecycle::Open,
            stream_epoch: None,
            sent: 0,
            parsed: 0,
            final_offset: None,
        }
    }

    pub(crate) fn run(&self) -> &RunKey {
        &self.run
    }

    pub(crate) fn process(&self) -> ProcessLifecycle {
        self.process
    }

    pub(crate) fn output(&self) -> OutputLifecycle {
        self.output
    }

    pub(crate) fn final_offset(&self) -> Option<WireU64> {
        self.final_offset.map(wire)
    }

    pub(crate) fn parsed_offset(&self) -> WireU64 {
        wire(self.parsed)
    }

    pub(crate) fn process_running(&mut self) -> Result<(), SafeError> {
        match self.process {
            ProcessLifecycle::Starting | ProcessLifecycle::Running => {
                self.process = ProcessLifecycle::Running;
                Ok(())
            }
            _ => Err(error("RUN_STATE_CONFLICT")),
        }
    }

    pub(crate) fn process_exited(&mut self) -> Result<(), SafeError> {
        match self.process {
            ProcessLifecycle::Starting | ProcessLifecycle::Running | ProcessLifecycle::Exited => {
                self.process = ProcessLifecycle::Exited;
                if self.output == OutputLifecycle::Open {
                    self.output = OutputLifecycle::Draining;
                }
                self.refresh_drained();
                Ok(())
            }
            ProcessLifecycle::Failed => Err(error("RUN_STATE_CONFLICT")),
        }
    }

    pub(crate) fn process_failed(&mut self) {
        self.process = ProcessLifecycle::Failed;
        if !matches!(
            self.output,
            OutputLifecycle::Degraded | OutputLifecycle::Incomplete
        ) {
            self.output = OutputLifecycle::Incomplete;
        }
    }

    pub(crate) fn output_started(&mut self, epoch: &str) -> Result<(), SafeError> {
        let epoch = WireU64::parse(epoch)?;
        if epoch.get() == 0 {
            return Err(SafeError::invalid("streamEpoch"));
        }
        match self.stream_epoch {
            None => {
                self.stream_epoch = Some(epoch);
                Ok(())
            }
            Some(current) if current == epoch => Ok(()),
            Some(_) => Err(error("STALE_OUTPUT_STREAM")),
        }
    }

    pub(crate) fn sent_through(&mut self, through: &str) -> Result<(), SafeError> {
        self.require_stream()?;
        let through = WireU64::parse(through)?.get();
        if through < self.sent {
            return Err(error("OUTPUT_OFFSET_BACKWARD"));
        }
        if self.final_offset.is_some() && through != self.sent {
            return Err(error("OUTPUT_ALREADY_ENDED"));
        }
        self.sent = through;
        Ok(())
    }

    pub(crate) fn output_end(&mut self, final_offset: &str) -> Result<(), SafeError> {
        let epoch = self.require_stream()?;
        self.output_end_for(&epoch.to_string(), final_offset)
    }

    pub(crate) fn output_end_for(
        &mut self,
        epoch: &str,
        final_offset: &str,
    ) -> Result<(), SafeError> {
        self.check_epoch(epoch)?;
        let final_offset = WireU64::parse(final_offset)?.get();
        if final_offset > self.sent {
            return Err(error("OUTPUT_END_BEYOND_SENT"));
        }
        if let Some(current) = self.final_offset {
            if current != final_offset {
                return Err(error("OUTPUT_END_CONFLICT"));
            }
        } else {
            self.final_offset = Some(final_offset);
        }
        if self.output == OutputLifecycle::Open {
            self.output = OutputLifecycle::Draining;
        }
        self.refresh_drained();
        Ok(())
    }

    pub(crate) fn parsed_through(&mut self, through: &str) -> Result<(), SafeError> {
        let epoch = self.require_stream()?;
        self.parsed_through_for(&epoch.to_string(), through)
    }

    pub(crate) fn parsed_through_for(
        &mut self,
        epoch: &str,
        through: &str,
    ) -> Result<(), SafeError> {
        self.check_epoch(epoch)?;
        let through = WireU64::parse(through)?.get();
        if through < self.parsed {
            return Err(error("OUTPUT_ACK_BACKWARD"));
        }
        if through > self.sent {
            return Err(error("OUTPUT_ACK_BEYOND_SENT"));
        }
        self.parsed = through;
        self.refresh_drained();
        Ok(())
    }

    pub(crate) fn mark_degraded(&mut self) -> Result<(), SafeError> {
        self.require_stream()?;
        self.output = OutputLifecycle::Degraded;
        Ok(())
    }

    pub(crate) fn mark_incomplete(&mut self) -> Result<(), SafeError> {
        self.output = OutputLifecycle::Incomplete;
        Ok(())
    }

    pub(crate) fn can_retire(&self) -> bool {
        self.process == ProcessLifecycle::Exited
            && matches!(
                self.output,
                OutputLifecycle::Drained | OutputLifecycle::Degraded | OutputLifecycle::Incomplete
            )
    }

    pub(crate) fn can_retire_as_complete(&self) -> bool {
        self.process == ProcessLifecycle::Exited && self.output == OutputLifecycle::Drained
    }

    fn require_stream(&self) -> Result<WireU64, SafeError> {
        self.stream_epoch
            .ok_or_else(|| error("OUTPUT_STREAM_NOT_READY"))
    }

    fn check_epoch(&self, epoch: &str) -> Result<(), SafeError> {
        let epoch = WireU64::parse(epoch)?;
        if self.stream_epoch != Some(epoch) {
            return Err(error("STALE_OUTPUT_STREAM"));
        }
        Ok(())
    }

    fn refresh_drained(&mut self) {
        if matches!(
            self.output,
            OutputLifecycle::Degraded | OutputLifecycle::Incomplete
        ) {
            return;
        }
        if self
            .final_offset
            .is_some_and(|final_offset| self.parsed == final_offset)
        {
            self.output = OutputLifecycle::Drained;
        } else if self.final_offset.is_some() || self.process == ProcessLifecycle::Exited {
            self.output = OutputLifecycle::Draining;
        }
    }
}

fn wire(value: u64) -> WireU64 {
    WireU64::parse(&value.to_string()).expect("internal lifecycle offsets are canonical")
}
